File: super-ugly-number.cc
// Super Ugly Number
typedef pair<int, int> pii;
#define REP(i, n) for (int i = 0; i < (n); i++)

class Solution {
public:
  int nthSuperUglyNumber(int n, vector<int>& primes) {
    int m = primes.size();
    vector<int> ugly(n, 1), idx(m, 0);
    priority_queue<pii, vector<pii>, greater<pii>> pq;
    REP(i, m)
      pq.emplace(primes[i], i);
    for (int i = 1; i < n; ) {
      pii x = pq.top();
      pq.pop();
      if (x.first != ugly[i-1])
        ugly[i++] = x.first;
      pq.emplace(ugly[++idx[x.second]] * primes[x.second], x.second);
    }
    return ugly[n-1];
  }
};

File: super-ugly-number.rs
// Super Ugly Number
impl Solution {
    pub fn nth_super_ugly_number(n: i32, primes: Vec<i32>) -> i32 {
        let m = primes.len();
        let mut a = vec![1];
        let mut ix = vec![0; m];
        while a.len() < n as usize {
            let mn = (0..m).fold(std::i32::MAX, |mn,i| mn.min(a[ix[i]]*primes[i]));
            a.push(mn);
            for i in 0..m {
                if a[ix[i]]*primes[i] == mn {
                    ix[i] += 1;
                }
            }
        }
        *a.last().unwrap()
    }
}
